rosbag2_cpp/src/rosbag2_cpp/cache/message_cache.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "rosbag2_cpp/cache/message_cache.hpp"
+#include "rosbag2_cpp/logging.hpp"
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+
+MessageCache::MessageCache(const uint64_t & max_buffer_size)
+{
+  primary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+  secondary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+}
+
+void MessageCache::push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg)
+{
+  // While pushing, we keep track of inserted and dropped messages as well
+  bool pushed = false;
+  {
+    std::lock_guard<std::mutex> cache_lock(cache_mutex_);
+    if (!flushing_) {
+      pushed = primary_buffer_->push(msg);
+    }
+  }
+  if (!pushed) {
+    messages_dropped_per_topic_[msg->topic_name]++;
+  }
+
+  notify_buffer_consumer();
+}
+
+void MessageCache::finalize()
+{
+  {
+    std::lock_guard<std::mutex> cache_lock(cache_mutex_);
+    flushing_ = true;
+  }
+  cache_condition_var_.notify_one();
+}
+
+void MessageCache::notify_buffer_consumer()
+{
+  {
+    std::lock_guard<std::mutex> cache_lock(cache_mutex_);
+    primary_buffer_can_be_swapped_ = true;
+  }
+  cache_condition_var_.notify_one();
+}
+
+void MessageCache::wait_for_buffer()
+{
+  std::unique_lock<std::mutex> lock(cache_mutex_);
+  if (!flushing_) {
+    // Required condition check to protect against spurious wakeups
+    cache_condition_var_.wait(
+      lock, [this] {
+        return primary_buffer_can_be_swapped_ || flushing_;
+      });
+    primary_buffer_can_be_swapped_ = false;
```

Comment:
what is variable for? I see that it's set to false here, and set to true in the notify function. But were is it actually being used?

Reply:
in the predicate for the conditional wait (two lines above the false assignment). This is due to the case of "spurious wake-ups" e.g. proper use of conditional variables should include checking whether the wake up was indeed due to intended logic change